core/templates/dev/head/services/StateTopAnswersStatsService.ts

```diff
@@ -25,75 +25,69 @@ require('services/ContextService.ts');
+     * @property {AnswerStats[]} answers - the collection of answers.
+     * @property {string} interactionId - the interaction id of the answers.
      */
 
-    /** @type {Object.<string, AnswerStatsCache[]>} */
-    var stateTopAnswersStatsCache = {};
-
     /** @type {boolean} */
     var isInitialized = false;
+    /** @type {!Object.<string, AnswerStatsCache>} */
+    var workingStateTopAnswersStats = {};
 
     /**
      * Updates the addressed info of all the answers cached for the given state
      * to reflect the state's current answer groups.
      * @param {string} stateName
      */
     var refreshAddressedInfo = function(stateName) {
-      var explorationId = ContextService.getExplorationId();
+      if (!workingStateTopAnswersStats.hasOwnProperty(stateName)) {
+        throw Error(stateName + ' does not exist.');
+      }
+
       var state = ExplorationStatesService.getState(stateName);
-      var interactionRulesService = $injector.get(
-        AngularNameService.getNameOfInteractionRulesService(
-          state.interaction.id));
-      var allAnswersCacheEntry =
-        stateTopAnswersStatsCache[stateName].allAnswers;
-      var unresolvedAnswersCacheEntry =
-        stateTopAnswersStatsCache[stateName].unresolvedAnswers;
-
-      // Clear the unresolvedAnswers array since many answers may now have
-      // different "addressed" values.
-      unresolvedAnswersCacheEntry.length = 0;
-
-      // Update the isAddressed data of each answer and put any unaddressed
-      // answers into the unresolvedAnswers array.
-      allAnswersCacheEntry.forEach(function(answerStats) {
-        answerStats.isAddressed =
+      var stateStats = workingStateTopAnswersStats[stateName];
+
+      if (stateStats.interactionId !== state.interaction.id) {
+        stateStats.answers.length = 0;
+        stateStats.interactionId = state.interaction.id;
+      }
+
+      // Update the isAddressed property of each answer.
+      var interactionRulesService = stateStats.interactionId === null ?
+        null :
+        $injector.get(AngularNameService.getNameOfInteractionRulesService(
+          stateStats.interactionId));
+      stateStats.answers.forEach(function(answerStats) {
+        answerStats.isAddressed = interactionRulesService !== null &&
           AnswerClassificationService.isClassifiedExplicitlyOrGoesToNewState(
-            stateName, state, answerStats.answer,
-            interactionRulesService);
-        if (!answerStats.isAddressed) {
-          unresolvedAnswersCacheEntry.push(answerStats);
-        }
+            stateName, state, answerStats.answer, interactionRulesService);
       });
     };
 
     var onStateAdded = function(stateName) {
-      stateTopAnswersStatsCache[stateName] = {
-        allAnswers: [],
-        unresolvedAnswers: []
-      };
+      var state = ExplorationStatesService.getState(stateName);
+      workingStateTopAnswersStats[stateName] =
+        {answers: [], interactionId: state.interaction.id};
```

Comment:
I wish we had a domain object for this, it's easy to make mistakes by forgetting to include a key...

But that can be a separate PR, we should get this in.

Reply:
Ack.